### PR TITLE
integration: add the basics of cross-integration concerns

### PR DESCRIPTION
**Pull request by freben**
The `ScmIntegrations` interface is similar to the `UrlReaders`, in the way that it is expected to be passed to any consumer that needs to perform SCM agnostic actions.

Examples of use cases I envision it will cover:
- I have a URL and I want to know the type of provider it refers to
- I have a web form where I want to list all integrations so the user can select one of them
- I have a URL to a catalog info file and I want to get the repo root from it
- I have a URL to a catalog info file and I want to get a browser friendly edit URL for it
- I want to parse a URL into its constituent parts (org/repo/path etc)

Mmmmmaybe?
- I want to decorate a request to this URL with auth

Possibly later also...?
- I want to know the default branch name for this repo
- I want to make a pull request in this repo
- I want to create a repo or check if it exists

### CODE REVIEW

File: packages/integration/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+   * Lists all registered integrations.
+   */
+  list(): ScmIntegration[];
+
+  /**
+   * Fetches a named integration
+   *
+   * @param name The name of a registered integration
+   */
+  byName(name: string): ScmIntegration | undefined;
```
Comment (Rugvip) [inline packages/integration/src/types.ts:41]: Perhaps `byHost` instead, along with renaming renaming `name` to `host` everywhere? I expected `name` to be something like `GitHub` or even a custom thing like `"GHE"` or `"Bogus BU's GitLab"`.

File: packages/integration/src/types.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+  /**
+   * Fetches an integration by URL
+   *
+   * @param name A URL that matches a registered integration
+   */
+  byUrl(url: string): ScmIntegration | undefined;
+};
+
+export type ScmIntegrationPredicateTuple = {
```
Comment (Rugvip) [inline packages/integration/src/types.ts:51]: Assuming this is intended to replace the types in `@backstage/backend-common` right?

File: .changeset/nine-forks-sleep.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/integration': minor
```
Comment (Rugvip) [inline .changeset/nine-forks-sleep.md:2]: Just `patch` right?

File: packages/integration/src/ScmIntegrationsImpl.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+import { GitHubIntegration } from './github/GitHubIntegration';
+import { GitLabIntegration } from './gitlab/GitLabIntegration';
+import {
+  ScmIntegration,
+  ScmIntegrationPredicateTuple,
+  ScmIntegrations,
+} from './types';
+
+export class ScmIntegrationsImpl implements ScmIntegrations {
+  static fromConfig(config: Config): ScmIntegrationsImpl {
```
Comment (Rugvip) [inline packages/integration/src/ScmIntegrationsImpl.ts:29]: This isn't being exported so no way to consume this atm 😅 

Perhaps make the `ScmIntegrations` interface `ScmIntegrationCollection` or something of the sort instead, and then make `ScmIntegrations` a class with just this single static `fromConfig` that is exported directly.

**Review by Rugvip — COMMENTED** [review]
Yup, overall 👍 , some name nits and export troubles

File: packages/integration/src/types.ts (hunk 1 above)
Comment (freben) [reply packages/integration/src/types.ts]: Removed this method. Also changed the type to have the fields `type` and `title` instead, making its purpose for end-user display (e.g. in a picker for where to scaffold your component). While there's no such field in the config yet, it is set to the hostname instead.

File: packages/integration/src/types.ts (hunk 2 above)
Comment (freben) [reply packages/integration/src/types.ts:52]: Which one, `UrlReaderPredicateTuple`?

File: .changeset/nine-forks-sleep.md (hunk 3 above)
Comment (freben) [reply .changeset/nine-forks-sleep.md]: Aight, changed

File: packages/integration/src/ScmIntegrationsImpl.ts (hunk 4 above)
Comment (freben) [reply packages/integration/src/ScmIntegrationsImpl.ts]: Added export and renamed as `ScmIntegrations implements ScmIntegrationRegistry`

**Review by Rugvip — APPROVED** [review]
:shipit:

**Pull request merged**